File: app/src/main/java/com/tunjid/androidbootstrap/viewholders/RouteItemViewHolder.java
package com.tunjid.androidbootstrap.viewholders;

import android.annotation.SuppressLint;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.tunjid.androidbootstrap.R;
import com.tunjid.androidbootstrap.adapters.RouteAdapter;
import com.tunjid.androidbootstrap.model.Route;
import com.tunjid.androidbootstrap.recyclerview.InteractiveViewHolder;

import androidx.core.graphics.drawable.DrawableCompat;
import androidx.core.widget.TextViewCompat;
import androidx.transition.AutoTransition;
import androidx.transition.TransitionManager;
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat;

public class RouteItemViewHolder extends InteractiveViewHolder<RouteAdapter.RouteAdapterListener>
        implements View.OnClickListener {

    private Route route;

    private TextView routeDestination;
    private TextView routeDescription;

    public RouteItemViewHolder(View itemView, RouteAdapter.RouteAdapterListener listener) {
        super(itemView, listener);

        routeDestination = itemView.findViewById(R.id.destination);
        routeDescription = itemView.findViewById(R.id.description);

        itemView.setOnClickListener(this);
        routeDescription.setOnClickListener(this);

        setIcons(true, routeDestination);
    }

    public void bind(Route route) {
        this.route = route;

        routeDestination.setText(route.getDestination());
        routeDescription.setText(route.getDescription());
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.description:
                adapterListener.onItemClicked(route);
                break;
            default:
                changeVisibility(routeDestination, routeDescription);
                break;
        }
    }

    @SuppressLint("ResourceAsColor")
    private void setIcons(boolean isDown, TextView... textViews) {
        int resVal = isDown ? R.drawable.anim_vect_down_to_right_arrow : R.drawable.anim_vect_right_to_down_arrow;

        for (TextView textView : textViews) {
            Drawable icon = AnimatedVectorDrawableCompat.create(itemView.getContext(), resVal);
            if (icon != null) {
                icon = DrawableCompat.wrap(icon.mutate());
                DrawableCompat.setTint(icon, R.color.dark_grey);
                DrawableCompat.setTintMode(icon, PorterDuff.Mode.SRC_IN);
                TextViewCompat.setCompoundDrawablesRelativeWithIntrinsicBounds(textView, null, null, icon, null);
            }
        }
    }

    private void changeVisibility(TextView clicked, View... changing) {
        TransitionManager.beginDelayedTransition((ViewGroup) itemView.getParent(), new AutoTransition());

        boolean visible = changing[0].getVisibility() == View.VISIBLE;

        setIcons(visible, clicked);

        AnimatedVectorDrawableCompat animatedDrawable = (AnimatedVectorDrawableCompat)
                TextViewCompat.getCompoundDrawablesRelative(clicked)[2];

        animatedDrawable.start();

        int visibility = visible ? View.GONE : View.VISIBLE;
        for (View view : changing) view.setVisibility(visibility);
    }
}

File: app/src/main/java/com/tunjid/androidx/viewholders/RouteItemViewHolder.kt
package com.tunjid.androidx.viewholders

import android.annotation.SuppressLint
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.widget.TextViewCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.AutoTransition
import androidx.transition.TransitionManager
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat
import com.tunjid.androidx.R
import com.tunjid.androidx.core.content.themeColorAt
import com.tunjid.androidx.core.graphics.drawable.withTint
import com.tunjid.androidx.model.RouteItem

class RouteItemViewHolder(
        itemView: View,
        private val delegate: (RouteItem.Destination) -> Unit
) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

    private var route: RouteItem? = null

    private val routeDestination: TextView = itemView.findViewById(R.id.destination)
    private val routeDescription: TextView = itemView.findViewById(R.id.description)

    init {
        itemView.setOnClickListener(this)
        routeDescription.setOnClickListener(this)

        setIcons(true, routeDestination)
    }

    fun bind(route: RouteItem) {
        this.route = route

        itemView.visibility = if (route is RouteItem.Destination) View.VISIBLE else View.INVISIBLE
        if (route !is RouteItem.Destination) return

        routeDestination.text = route.destination
        routeDescription.text = route.description
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.description -> (route as? RouteItem.Destination)?.let(delegate)
            else -> changeVisibility(routeDestination, routeDescription)
        }
    }

    @SuppressLint("ResourceAsColor")
    private fun setIcons(isDown: Boolean, vararg textViews: TextView) {
        val context = itemView.context
        val resVal = if (isDown) R.drawable.anim_vect_down_to_right_arrow else R.drawable.anim_vect_right_to_down_arrow

        for (textView in textViews) {
            val icon: Drawable? = AnimatedVectorDrawableCompat.create(itemView.context, resVal)?.withTint(context.themeColorAt(R.attr.prominent_text_color))
            if (icon != null) {
                DrawableCompat.setTintMode(icon, PorterDuff.Mode.SRC_IN)
                TextViewCompat.setCompoundDrawablesRelativeWithIntrinsicBounds(textView, null, null, icon, null)
            }
        }
    }

    private fun changeVisibility(clicked: TextView, vararg changing: View) {
        TransitionManager.beginDelayedTransition(itemView.parent as ViewGroup, AutoTransition())

        val visible = changing[0].visibility == View.VISIBLE

        setIcons(visible, clicked)

        val animatedDrawable = TextViewCompat.getCompoundDrawablesRelative(clicked)[2] as AnimatedVectorDrawableCompat

        animatedDrawable.start()

        val visibility = if (visible) View.GONE else View.VISIBLE
        for (view in changing) view.visibility = visibility
    }
}
